Execution environment test builds

oraNod opened the pull request:
Addresses #16 

Opening as a draft to start a conversation and get some advice on how to proceed.

- At the moment I've been using an arg parser to pass things like the path and collection namespace and name. This has been a convenient way to test things and work through the logic. I plan to modify things and get this better integrated with the rest of the project but wanted to open a PR and work through all that collaboratively.
- Need to add docstrings and type hints.
- Need to add actual tests for ee builds. Should I port over the contents of https://github.com/ansible-collections/community.docker/tree/main/tests/ee ?

felixfontein commented on src/antsibull_nox/ee.py:
```diff
@@ -0,0 +1,161 @@
+from pathlib import Path
+from typing import Optional
+
+import yaml
```
Instead of importing `yaml`, better use `from antsibull_fileutils.yaml import load_yaml_file, store_yaml_file` and call these functions.

felixfontein commented on src/antsibull_nox/ee.py:
```diff
@@ -0,0 +1,161 @@
+    return parser.parse_args()
+
+
+def check_galaxy_version(args: Args):
```
Check out how `antsibull_nox.sessions.build_import_check`'s `build_import_check()` does this. Also this function should receive a path as its argument, and not the full `Args` object.

oraNod replied:
Thanks for the pointer, @felixfontein I've been looking at it but would like to clarify my understanding please.

`build_import_check()` creates temporary files instead of editing the actual `galaxy.yml` right? Would it be better to import this like `from .sessions.build_import_check import add_build_import_check` or rewrite my `check_galaxy_version()` to behave in a similar fashion?

felixfontein replied:
> `build_import_check()` creates temporary files instead of editing the actual `galaxy.yml` right?

Yes. It creates a copy of the collection (`copy_collection(Path.cwd(), collection_dir)`) and then modifies galaxy.yml in there if necessary.

> Would it be better to import this like `from .sessions.build_import_check import add_build_import_check` or rewrite my `check_galaxy_version()` to behave in a similar fashion?

I'd simply copy the required code over for now. We can refactor it later so there's a common functionality.

felixfontein replied:
I extracted the code in #83.

oraNod replied:
Ah, awesome. Thanks @felixfontein ! I appreciate your patience while I work through this. Just to let you know I'm going to mostly be away until early next week but will make finishing this off a priority when I get back. It has been a good experience getting to integrate some code into your project so cheers again for that.

oraNod commented:
@felixfontein Hey, thanks for #83 that looks like it simplifies things. I can just import with this:

```
from .collection import (
    build_collection,
)
```

Is that correct? I'm just wondering though, will that result in the collection being built for each session? On the one hand that makes sense if you're running individual test sessions (not sure if that's even possible). But if you're running all test sessions at once, does that mean the collection gets built multiple times?

felixfontein commented:
> Is that correct?

Yes.

> I'm just wondering though, will that result in the collection being built for each session?

Yes.

> On the one hand that makes sense if you're running individual test sessions (not sure if that's even possible). But if you're running all test sessions at once, does that mean the collection gets built multiple times?

That's true, but it could also happen in different contexts (i.e. with different ansible-core versions or branches). Also building is usually the fastest part of all tests involving builds, so I think it's ok. (At least for now; it can always be improved later on.)

oraNod commented:
@felixfontein Hey, thanks again for your patience as I work through this PR. I feel like I've made some decent progress and have removed the args that I was using and have added a noxfile to a collection and calling the ee test build from there. It seems to be working as expected, generating the ee.yml files and building the images with the collection.

At this point there are two main pieces I think I need to tackle:

- Add a parameter to specify a playbook to run within the ee and the ability to execute that.
- Integrate into the session loader and TOML schema.

I plan to continue working on this over the next week or so. I'd definitely welcome any feedback in the meantime, especially if you have any thoughts or pointers on the above two things. Cheers.

oraNod commented on src/antsibull_nox/sessions/execution_environment.py:
```diff
@@ -0,0 +1,143 @@
+        - collection_tarball_path: Path to the built collection tarball
+        - built_image_names: List of built container images
+    """
+    collection_tarball_result = build_collection(session)
```
So I'm passing in collection data and then calling a function that builds the same data again?

![image](https://github.com/user-attachments/assets/95ab21ed-5311-4825-9ad4-ddd7661844c9)

oraNod commented on src/antsibull_nox/sessions/execution_environment.py:
```diff
@@ -0,0 +1,143 @@
+
+def add_execution_environment_session(
+    sessions_dict: dict,
+    collection_data: CollectionData,
```
Thanks for spotting this @felixfontein I've removed `collection_data` here.

felixfontein commented:
It's probably better to not have `container_engine` as part of the noxfile, since that's something that really depends on the system, and not on the collection. I'd probably simply set it to a default for the first version and allow to override it with an environment variable; the next step would be to change the default to some auto-detection (look for `docker` and `podman` CLI tools on path; ansible-test does that as well I think).

oraNod commented:
Rebased to squash a few trivial commits and clean up the history a little.

felixfontein commented on pyproject.toml:
```diff
@@ -84,6 +84,7 @@ typing = [
     "antsibull-docutils >= 1.2.0",
     "types-docutils",
     "types-PyYAML",
+    "types-jinja2",
```
This shouldn't be needed anymore:
```suggestion
```

felixfontein commented:
> @felixfontein I've been looking through ansible-test to understand. I see `DockerCommand.detect()` here: https://github.com/ansible/ansible/blob/8c3d23513fb3eff4fae53fda8ee674f1ceca587f/test/lib/ansible_test/_internal/docker_util.py#L451
> 
> Looks like docker is tried first and then podman (but you can change that with another environment variable named `ANSIBLE_TEST_PREFER_PODMAN`) . There is some other logic in the detection that I don't fully understand (like [here](https://github.com/ansible/ansible/blob/8c3d23513fb3eff4fae53fda8ee674f1ceca587f/test/lib/ansible_test/_internal/docker_util.py#L464)) and some fancy stuff like caching.

The line you don't understand is checking for podman's docker compatibility shim, where basically `docker` is an alias for `podman`. If you run `docker -v` with that shim active, you get output like `podman version 5.5.2` (which is exactly what `podman -v` outputs). In that case it's better to use `podman` right away instead of `docker`. (Without the shim, you get something like `Docker version 28.3.0, build 38b7060a21`.)

> Anyway I think we could do something to mimic ansible-test by using subprocess to capture the output of `<engine> -v`. We could also change it so docker is preferred, like in ansible-test.

Some auto-detection would be great, but for the first verison something even simpler (always use `x`, and allow to configure it via environment variable) is totally fine. That's up to your discretion, like how much time you have and how much work you want to put into it :) (I can add the missing pieces later, if there are any.)

> I wonder if it is better to keep the environment variable from my previous commit so you can explicitly set the container engine with `CONTAINER_ENGINE` or if consistency rules and we should create a different variable named something like `CONTAINER_ENGINE_PREFER_PODMAN`?

I think `CONTAINER_ENGINE` is fine for now, with values `podman` and `docker`. We can later add values like `auto`, `auto-prefer-docker`, `auto-prefer-podman`, to allow users to express exactly what they want.

> I also wonder if this is something that would be better as a separate module. If so, maybe it's better to leave this as an environment variable for now, like [your comment](https://github.com/ansible-community/antsibull-nox/pull/69#issuecomment-3085375455).

At some point it would make sense to move it out, so it can also used for other tests that need to choose a container engine (for `ansible-test` `ANSIBLE_TEST_PREFER_PODMAN` could be set appropriately if it hasn't been set already and if `CONTAINER_ENGINE` is set to something else than `auto`). But again, this is something that can also easily be done later on...

felixfontein commented on src/antsibull_nox/sessions/execution_environment.py:
```diff
@@ -0,0 +1,275 @@
+            f" using {container_engine}"
+        )
+
+        session.run("/bin/sh", "-c", "pwd && ls -lah")
```
This looks like debug output to me; if it is, it should be removed before merging:
```suggestion
```

oraNod commented:
@felixfontein I've added some functions to `src/antsibull_nox/ee_config.py` so you can now set custom ee config in the project noxfile. I figured there would be two ways to do that. One way is to use an `execution-environment.yml` file in the collection. The other way is to specify certain config fields.

This could probably use some improvement but in noxfile.py I've added changes so you can now...

### Specify config fields

```
from antsibull_nox.ee_config import create_ee_config
from antsibull_nox.sessions.execution_environment import ExecutionEnvironmentData, add_execution_environment_sessions

my_config = create_ee_config (
    base_image="fedora:41",
    dependencies={"ansible_core": {"package_pip": "https://github.com/ansible/ansible/archive/stable-2.16.tar.gz"}}
)

my_ee = ExecutionEnvironmentData(
    name="my_nox_test_ee",
    description="My antsibull-nox test ee",
    config=my_config,
    test_playbooks=["nora_playbook.yml"]
)

add_execution_environment_sessions(execution_environments=[my_ee])
```

### Specify an ee config file

```
from antsibull_nox.ee_config import load_ee_config
from antsibull_nox.sessions.execution_environment import ExecutionEnvironmentData, add_execution_environment_sessions

my_config = load_ee_config("execution-environment.yml")
my_ee = ExecutionEnvironmentData(
    name="my_nox_test_ee_file",
    description="My antsibull-nox test ee",
    config=my_config,
    test_playbooks=["nora_playbook.yml"]
)

add_execution_environment_sessions(execution_environments=[my_ee])
```

felixfontein commented:
I'm not sure `load_ee_config` should be there at all. The only relevant config file should be `antsibull-nox.toml`, where support for the new session needs to be added at some point (either in this PR or a follow-up PR).

> @felixfontein Besides whatever nits and tweaks there might be to the commits I've added, is there anything else you think is required beyond test coverage and docs? I can work on adding those this week.

A changelog fragment is definitely needed :)

Regarding test coverage and docs:
- For test coverage, add some basic EE test to the test collection (in `tests/test-collection/`).
- Documentation should happen once support to `antsibull-nox.toml` is added, so only add something to the docs in this PR if you add support for `antsibull-nox.toml` first.

felixfontein commented:
> > I'm not sure `load_ee_config` should be there at all. The only relevant config file should be `antsibull-nox.toml`, where support for the new session needs to be added at some point (either in this PR or a follow-up PR).
> 
> @felixfontein Can you explain this a little more please? I added `load_ee_config` with the idea that collections might have an `execution-environment.yml` file in the project. This allows you to pass that file instead of adding EE config settings to the noxfile.

1. You can also simply `from antsibull_fileutils.yaml import load_yaml_file` and then use `load_yaml_file` directly to load a YAML file's contents. Also adding that API now means we have to support it for a very long time, and if there is not a very clear benefit of having it, I think it's better to leave it away. (Especially since it's so easy to directly load YAML content.)
2. Adding sessions to `noxfile.py` is something that most collections should not do / have no use for. So if we want to support this, we should add this as meta-data to `ExecutionEnvironmentData`. But that's - again - special extensions (used by very few collections) that can be added later, and should not be part of the first basic implementation IMO.

oraNod commented:
> > > I'm not sure `load_ee_config` should be there at all. The only relevant config file should be `antsibull-nox.toml`, where support for the new session needs to be added at some point (either in this PR or a follow-up PR).
> > 
> > 
> > @felixfontein Can you explain this a little more please? I added `load_ee_config` with the idea that collections might have an `execution-environment.yml` file in the project. This allows you to pass that file instead of adding EE config settings to the noxfile.
> 
>     1. You can also simply `from antsibull_fileutils.yaml import load_yaml_file` and then use `load_yaml_file` directly to load a YAML file's contents. Also adding that API now means we have to support it for a very long time, and if there is not a very clear benefit of having it, I think it's better to leave it away. (Especially since it's so easy to directly load YAML content.)
> 
>     2. Adding sessions to `noxfile.py` is something that most collections should not do / have no use for. So if we want to support this, we should add this as meta-data to `ExecutionEnvironmentData`. But that's - again - special extensions (used by very few collections) that can be added later, and should not be part of the first basic implementation IMO.

Thanks @felixfontein I've probably been getting distracted by how I've been testing this stuff. I'll try and get to this again today and will update. Cheers.

felixfontein commented on src/antsibull_nox/ee_config.py:
```diff
@@ -0,0 +1,104 @@
+    if not isinstance(config["additional_build_files"], list):
+        raise ValueError(
+            f"Expected a list at additional_build_files,"
+            f" but found {type(config["additional_build_files"])}"
```
Older Python versions (like 3.9) do not like it if you're using the same quotation marks inside `{}` that are used for the f-string itself. (That's why the unit tests are failing for Python 3.9, 3.10, and 3.11.)
```suggestion
            f" but found {type(config['additional_build_files'])}"
```

oraNod replied:
Thanks @felixfontein

felixfontein commented:
> 1. I see a lint error: `src/antsibull_nox/interpret_config.py:131:0: R0912: Too many branches (13/12) (too-many-branches)`
> 
> Should we suppress this one for now?

Yep, simply suppress it, or bump `max-branches=12` to `max-branches=13` in `.pylintrc.automated`.

I'll try to reproduce the podman issue, but I'm not sure when I'll manage to do that (definitely not today :) )

felixfontein commented on src/antsibull_nox/ee_config.py:
```diff
@@ -0,0 +1,104 @@
+Execution environment definition generator.
+"""
+
+import typing as t
```
```suggestion
from __future__ import annotations

import typing as t
```

oraNod commented:
> > 1. I see a lint error: `src/antsibull_nox/interpret_config.py:131:0: R0912: Too many branches (13/12) (too-many-branches)`
> > 
> > Should we suppress this one for now?
> 
> Yep, simply suppress it, or bump `max-branches=12` to `max-branches=13` in `.pylintrc.automated`.
> 
> I'll try to reproduce the podman issue, but I'm not sure when I'll manage to do that (definitely not today :) )

Cheers. I failed to point out earlier that I've noticed that podman issue before. I'm not entirely sure what's going on with that but it appears to be just a podman issue. The ee-check session works just fine with docker. So maybe we can leave it as a follow on?

I still need to create that issue to follow up on the container engine auto detection. As part of that maybe we just need to add a bit of logic for podman to run outside nox's session directory.

felixfontein commented:
> You should probably use `antsibull_nox.paths.create_temp_directory()` to create another temp directory and set `TMPDIR` for the ansible-navigator call to that directory...

The docstring there was slightly wrong; in your specific setup (and mine) it would have worked, but it might not in general. In any case, there's now a better function: https://github.com/ansible-community/antsibull-nox/blob/main/src/antsibull_nox/paths.py#L194 Setting its result as `TMPDIR` for the ansible-navigator invocation should fix the podman problems.

oraNod commented on src/antsibull_nox/interpret_config.py:
```diff
@@ -304,6 +308,24 @@ def _add_sessions(sessions: Sessions) -> None:
+    if sessions.ee_check:
+        execution_environments = []
+
+        for ee_config in sessions.ee_check.execution_environments.values():
```
I guess an alternative approach here might be to do `for key in sessions.ee_check.execution_environments:` but I found that `values()` just does the job. Maybe it is less explicit and not as clear though. Is there any preference?

oraNod replied:
Well... now that you _say_ it like that, yes. Yes that is a good hint. Thanks @felixfontein

oraNod replied:
Wait. The change to a list restores the same problem I encountered before and that Pydantic didn't like. With this change I see:

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for Config
sessions.ee_check.execution_environments
  Input should be a valid list [type=list_type, input_value={'fedora-toolbox': {'name...fedora-toolbox:latest'}}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.11/v/list_type
```

felixfontein replied:
Pydantic complains that you are providing a dictionary, while you need to provide a list.

felixfontein replied:
Yep, you need an array of tables. You can find similar examples in the existing docs btw ;-) This section has a few: https://ansible.readthedocs.io/projects/antsibull-nox/config-file/#extra-checks-action-groups-unwanted-files-trailing-whitespace-unwanted-charactersregular-expressions

oraNod replied:
Ha! Yeah I actually started by combing through the docs and noticed some of those examples. This was a good one to learn.

felixfontein commented on src/antsibull_nox/config.py:
```diff
@@ -173,6 +175,63 @@ class SessionLicenseCheck(_BaseModel):
+
+    default: bool = False
+
+    execution_environments: dict[str, ExecutionEnvironmentConfig] = {}
```
I would simply use a list here. Also I would remove the default, since not specifying anything doesn't really make sense.
```suggestion
    execution_environments: list[ExecutionEnvironmentConfig]
```

felixfontein commented on src/antsibull_nox/config.py:
```diff
@@ -173,6 +175,63 @@ class SessionLicenseCheck(_BaseModel):
+    name: str
+    description: t.Optional[str] = None
+    test_playbooks: list[str]
+    version: int = 3
```
I'd restrict the possible values of `version` to `3`, since the code currently doesn't allow anything else:
```suggestion
    version: t.Literal[3] = 3
```

felixfontein commented on src/antsibull_nox/sessions/ee_check.py:
```diff
@@ -0,0 +1,282 @@
+        )
+
+        playbook_dir = Path.cwd()
+        temp_dir = create_temp_directory("ansible-navigator-")
```
You need
```suggestion
        temp_dir = get_outside_temp_directory(playbook_dir.absolute())
```
here (see https://github.com/ansible-community/antsibull-nox/pull/69#issuecomment-3101405063).

felixfontein commented on src/antsibull_nox/sessions/ee_check.py:
```diff
@@ -0,0 +1,282 @@
+        temp_dir = create_temp_directory("ansible-navigator-")
+
+        for playbook in execution_environment.test_playbooks:
+            playbook_path = playbook_dir / playbook
```
```suggestion
```

felixfontein commented on src/antsibull_nox/config.py:
```diff
@@ -173,6 +175,63 @@ class SessionLicenseCheck(_BaseModel):
+    test_playbooks: list[str]
+    version: t.Literal[3] = 3
+    base_image_name: str
+    ansible_core_source: t.Literal["package_pip", "package_system"] = "package_pip"
```
Maybe it would be easier to not provide such fine-grained configuration for the EE, but simply allow the user to provide the EE config. (Unfortunately TOML is not the greatest format to do that...)

felixfontein commented on .gitignore:
```diff
@@ -134,3 +134,6 @@ dmypy.json
 *.build
 *.deps
+
+# Ansible stuff
+.ansible/
```
This can be more specific:
```suggestion
# VSCode Ansible extension
/.ansible/
```

felixfontein commented on changelogs/fragments/16-ee-check.yml:
```diff
@@ -0,0 +1,2 @@
+major_changes:
```
We generally use `minor_changes` for new features:
```suggestion
minor_changes:
```

felixfontein commented on docs/config-file.md:
```diff
@@ -1143,3 +1143,51 @@ It simply runs `ansible-lint`.
+## Execution environment check
+
+antsibull-nox allows you to test your collection against an execution environment (EE).
+The `ee-check` session is added with the `[[sessions.ee_check.execution_environments]]` section in `antsibull-nox.toml`.
```
```suggestion
The `ee-check` session is added with the `[[sessions.ee_check.execution_environments]]` sections (one for every EE) in `antsibull-nox.toml`.
```

felixfontein commented on docs/config-file.md:
```diff
@@ -1143,3 +1143,51 @@ It simply runs `ansible-lint`.
+The `ee-check` session is added with the `[[sessions.ee_check.execution_environments]]` section in `antsibull-nox.toml`.
+It accepts the following options:
+
+* `name: str` (Required field): Specifies a unique name for the `ee-check` session.
```
```suggestion
* `name: str` (**required**): Specifies a unique name for the `ee-check` session.
```
Same for the other required ones.

felixfontein commented on docs/config-file.md:
```diff
@@ -1143,3 +1143,51 @@ It simply runs `ansible-lint`.
+* `test_playbooks: list[str]` (Required field): Specifies a list of playbooks that test the collection against the EE.
+* `version: t.Literal[3]` (default `3`): Configures the schema version for the EE definition.
+* `base_image_name: str` (default `"registry.fedoraproject.org/fedora-toolbox:latest"`): Specifies the base image to use when building the EE.
+* `ansible_core_source: t.Literal["package_pip", "package_system"]` (default `"package_pip"`): Configures the source for installing the `ansible-core` package.
```
```suggestion
* `ansible_core_source: "package_pip" | "package_system"` (default `"package_pip"`): Configures the source for installing the `ansible-core` package.
```

felixfontein commented on docs/config-file.md:
```diff
@@ -1143,3 +1143,51 @@ It simply runs `ansible-lint`.
+It accepts the following options:
+
+* `name: str` (Required field): Specifies a unique name for the `ee-check` session.
+* `description: t.Optional[str]` (default `None`): Specifies a description for the `ee-check` session.
```
```suggestion
* `description: str | None` (default `None`): Specifies a description for the `ee-check` session.
```

felixfontein commented on docs/config-file.md:
```diff
@@ -1143,3 +1143,51 @@ It simply runs `ansible-lint`.
+* `ansible_runner_package: str` (default `"ansible-runner"`): Specifies the name of the `ansible-runner` package.
+* `system_packages: list[str]` (default `"[]"`): Specifies a list of system packages to build into the EE.
+* `python_packages: list[str]` (default `"[]"`): Specifies a list of Python packages to build into the EE.
+* `python_interpreter_package: t.Optional[str]`(default `None`): Defines the Python system package name for the EE.
+* `python_path: t.Optional[str]`(default `None`): Specifies the path to the Python interpreter.
```
```suggestion
* `python_interpreter_package: str | None`(default `None`): Defines the Python system package name for the EE.
* `python_path: str | None`(default `None`): Specifies the path to the Python interpreter.
```

felixfontein commented on src/antsibull_nox/sessions/ee_check.py:
```diff
@@ -0,0 +1,280 @@
+        "dependencies": {
+            "ansible_core": {
+                "package_pip": "https://github.com/ansible/ansible/archive/devel.tar.gz",
+            },
+            "ansible_runner": {
+                "package_pip": "ansible-runner",
+            },
+            "python_interpreter": {
+                "package_system": "python3.11 python3.11-pip "
+                "python3.11-wheel python3.11-cryptography",
+                "python_path": "/usr/bin/python3.11",
+            },
+        },
+    },
+    test_playbooks=["tests/ee/all.yml"],
+)
+
+EXAMPLE_EE_DATA_2 = ExecutionEnvironmentData(
+    name="2.15-rocky-9",
+    description="ansible-core 2.15 @ Rocky Linux 9",
+    config={
+        "version": 3,
+        "images": {
+            "base_image": {
+                "name": "quay.io/rockylinux/rockylinux:9",
+            },
+        },
+        "dependencies": {
+            "ansible_core": {
+                "package_pip": "https://github.com/ansible/ansible/archive/stable-2.15.tar.gz",
+            },
+            "ansible_runner": {
+                "package_pip": "ansible-runner",
+            },
+        },
+    },
+    test_playbooks=["tests/ee/all.yml"],
+)
+
+
```
I think these example EEs can be removed:
```suggestion
```

felixfontein commented on tests/test-collection/antsibull-nox.toml:
```diff
@@ -6,3 +6,20 @@
+[[sessions.ee_check.execution_environments]]
+name = "fedora-toolbox"
+description = "Testing EE builds with the fedora toolbox"
+test_playbooks = ["tests/test-collection/ee.yml"]
```
This won't work, since `ee.yml` is in the collection root and not in a subdirectory `tests/test-collection` of the collection root (which would be `tests/test-collection/tests/test-collection/` in the repository):
```suggestion
test_playbooks = ["ee.yml"]
```

felixfontein commented on tests/test-collection/antsibull-nox.toml:
```diff
@@ -6,3 +6,20 @@
 [sessions.lint]
 run_mypy = false
+
+[[sessions.ee_check.execution_environments]]
```
If the EE sessions should actually run by default, you should mark them with `default = true`:
```suggestion
[sessions.ee_check]
default = true

[[sessions.ee_check.execution_environments]]
```
(This is also not documented in docs/config-file.md right now.)

oraNod commented on tests/test-collection/antsibull-nox.toml:
```diff
@@ -6,3 +6,23 @@
+
+[[sessions.ee_check.execution_environments]]
+name = "minimal"
+test_playbooks = ["tests/test-collection/ee.yml"]
```
```suggestion
test_playbooks = ["ee.yml"]
```

felixfontein commented on tests/test-collection/galaxy.yml:
```diff
@@ -4,5 +4,11 @@
+version: 1.0.0
+readme: README.md
+authors:
+  - Felix Fontein felix@fontein.de
```
I think just the name is also fine :)
```suggestion
  - Felix Fontein
```

felixfontein commented on tests/test-collection/antsibull-nox.toml:
```diff
@@ -6,3 +6,23 @@
 run_mypy = false
+
+[sessions.ee_check]
+default = true
```
This isn't documented yet.

felixfontein commented on src/antsibull_nox/config.py:
```diff
@@ -173,6 +175,63 @@ class SessionLicenseCheck(_BaseModel):
+    description: t.Optional[str] = None
+    test_playbooks: list[str]
+    version: t.Literal[3] = 3
+    base_image_name: str = "registry.fedoraproject.org/fedora-toolbox:latest"
+    ansible_core_source: t.Literal["package_pip", "package_system"] = "package_pip"
+    ansible_core_package: str = "ansible-core"
+    ansible_runner_package: str = "ansible-runner"
+    system_packages: list[str] = []
+    python_packages: list[str] = []
+    python_interpreter_package: t.Optional[str] = None
+    python_path: t.Optional[str] = None
```
How about simplifying this to:
```suggestion
    config: dict[str, t.Any] = {}
```
Then you can specify anything in the EE definition explicitly:
```toml
[[sessions.ee_check.execution_environments]]
name = "fedora-toolbox"
description = "Testing EE builds with the fedora toolbox"
test_playbooks = ["ee.yml"]
config.images.base_image.name = "registry.fedoraproject.org/fedora-toolbox:latest"
config.dependencies.ansible_core.package_pip = "ansible-core"
config.dependencies.ansible_runner.package_pip = "ansible-runner"
config.dependencies.system = [
    "git",
    "curl",
]
config.dependencies.python = [
    "jinja2",
    "pyyaml",
    "requests",
]
config.dependencies.python_interpreter.package_system = "python3"
config.dependencies.python_interpreter.python_path = "/usr/bin/python3"
```
